### DISCUSSION TITLE
Add SA for every new connection

### DISCUSSION THREAD

**Discussion by raocool** [strongswan/strongswan#1885] [Q&A]
I have a server that accepts connection. I would like to configure charon such that a new SA is added for every new connection. I know the two IP addresses and the listener port but not the incoming port. Is such a configuration  possible and if so how? 

Thanks much.
Shoaib

**Comment by tobiasbrunner** [comment]
Sorry, I don't understand what you want to achieve or what the difference would be to how things work by default.

**Comment by raocool** [comment]
We have HW that works better with individual SA's for each connection instead of using one SA between two systems. So I am looking at a policy configuration which adds a connection-specific SA. So when the SYN arrives it matches a wildcard policy and generates a trap, which results in a specific SA being added to the kernel. 
I have tried adding a policy via setsockopt on the listener, which does create an ACQUIRE but charon does not know what to do and hence my question.
Thanks for your response, if the requirement is still not clear please let me know. I am also willing to contribute code if we agree on an enhancement.
Regards,
Shoaib
    On Tuesday, September 5, 2023 at 12:52:07 AM PDT, Tobias Brunner ***@***.***> wrote:  
 
 


Sorry, I don't understand what you want to achieve or what the difference would be to how things work by default.

**Reply by tobiasbrunner** [reply]
> We have HW that works better with individual SA's for each connection instead of using one SA between two systems.

Why is that?

> So I am looking at a policy configuration which adds a connection-specific SA. So when the SYN arrives it matches a wildcard policy and generates a trap, which results in a specific SA being added to the kernel. 

That's currently not possible as the kernel doesn't support PFP ("populate from packet") as it only sends an acquire for the first 4-tuple until that acquire times out or results in an actual SA. And strongSwan, in turn, doesn't support narrowing to the packet-specific traffic selector from the acquire either.

**Comment by raocool** [comment]
What if I were to submit a patch? Is there a possibility that strongswan would accept it provided the Linux kernel accepts the change?
Thanks,
Shoaib

    On Wednesday, September 6, 2023 at 12:59:36 AM PDT, Tobias Brunner ***@***.***> wrote:  
 
 



We have HW that works better with individual SA's for each connection instead of using one SA between two systems.


Why is that?


So I am looking at a policy configuration which adds a connection-specific SA. So when the SYN arrives it matches a wildcard policy and generates a trap, which results in a specific SA being added to the kernel.


That's currently not possible as the kernel doesn't support PFP ("populate from packet") as it only sends an acquire for the first 4-tuple until that acquire times out or results in an actual SA. And strongSwan, in turn, doesn't support narrowing to the packet-specific traffic selector from the acquire either.

**Reply by tobiasbrunner** [reply]
A possibility? Sure. But obviously depends on the details.
